src/metabase/models/params.clj

```diff
@@ -203,26 +206,112 @@
   [cards]
   (reduce set/union #{} (map card->template-tag-field-ids cards)))
 
+;; Following function is dynamic for testing purposes.
+(defn ^:dynamic *filterable-columns-for-query*
+  "Get filterable columns for query."
+  [database-id dataset-query]
+  (-> (lib/query (lib.metadata.jvm/application-database-metadata-provider database-id)
+                 dataset-query)
+      (lib/filterable-columns)))
+
+(defn- ensure-filterable-columns-for-card
+  [ctx card]
+  (if (contains? (get ctx :card-id->filterable-columns) (:id card))
+    ctx
+    (let [database-id   (:database_id card)
+          dataset-query (:dataset_query card)]
+      (if-not (and (not-empty dataset-query) (pos-int? database-id))
+        ctx
+        (update ctx :card-id->filterable-columns assoc (:id card)
+                (*filterable-columns-for-query* database-id dataset-query))))))
+
+(defn- field-id-from-dashcards-filterable-columns
+  [ctx param-dashcard-info]
+  (let [param-target       (get-in param-dashcard-info [:parameter :target])
+        card-id            (get-in param-dashcard-info [:dashcard :card :id])
+        filterable-columns (get-in ctx [:card-id->filterable-columns card-id])]
+    (if-some [field-id (lib.util.match/match-one param-target
+                         [:field (field-name :guard string?) _]
```

Comment:
> Are we concerned here about :expression or :aggregation refs? 

To the best of my knowledge, no. That's the original logic _moved_ from what was `dashcards->param-field-ids` in the original PR, or the `param-target->field-id` used in the PR https://github.com/metabase/metabase/pull/45068 that's a base for this PR.

> Why not use lib.equality functions like find-matching-column?

I could, but in this PR I've aimed to add context for preserving the `filterable-columns` results in between computations instead of modifying original logic.